### ISSUE TITLE
Dependency resolution should taken into account the default registry 

### ISSUE THREAD

**Issue by jbw976** [crossplane/crossplane#5270]
### What happened?

@turkenh noticed while reviewing https://github.com/crossplane/crossplane/pull/5261#pullrequestreview-1834659686 that when the package manager performs dependency resolution, it does not take into account the default registry.

This has been a problem for awhile now, but will likely be exacerbated by the default registry changing in v1.15.

### How can we reproduce it?

@turkenh has provided the full scenario in https://github.com/crossplane/crossplane/pull/5261#pullrequestreview-1834659686, see details there.

### What environment did it happen in?
Crossplane version: found/tested with the code changes in #5261, but a similar scenario should exist in v1.14 too.

**Comment by phisco** [comment]
We discussed it with @ezgidemirel and agreed this is indeed not a bug, but intended behaviour.

Quoting from https://github.com/crossplane/crossplane/pull/2009: `All dependencies and revisions will be added to the Lock in the same format they were declared.`

We both agree with the above statement, which means that registry defaulting logic should be kept outside of the Lock file as this ensures its consistency.

Let's take as an example an enterprise user, building their own packages depending on other packages they've built and other external packages built by Upbound or other users. Assuming their packages are pushed to their own registry, instead of `xpkg.upbound.io`, so they set the default registry to that (just to be sure). 

External packages will mostly have fully qualified dependencies, because that's what we encourage and it makes sense to do so.

For their own packages they might want to use relative references or fully qualified ones explicitly specifying their registry. As long as they are consistent, everything will work as expected. The only issue would arise if they decided to mix the two approaches.

What @turkenh described [here](https://github.com/crossplane/crossplane/pull/5261#pullrequestreview-1834659686) falls under this mixed references scenario: the `Provider` was previously installed either directly or as a dependency using its full name reference, `xpkg.upbound.io/crossplane-contrib/provider-helm:v0.16.0`, but then the `Function` specified it as a dependency using a relative path, `crossplane-contrib/provider-helm`.

Actually, everything would have worked just fine if the `Provider` above had been installed using a relative reference in the first place (assuming the default registry served it, either because the image was pushed to their registry or they are just proxying `xpkg.upbound.io`). Obviously, as soon as one starts to use external packages with fully specified dependencies, this becomes an issue. However, that's what https://github.com/crossplane/crossplane/issues/4299 is about, and we can consider how to handle it at the lock file level once we get to it.

So, to sum it up, as long as a user is consistent in their usage of fully or relative references, everything will work fine. We should continue encouraging people to use explicit references and maybe improve the story around migrations if needed. But we think we should continue with our assumption that the lock file only takes into consideration exactly what has been specified, and not runtime configuration such as the default registry, which might change over time.

**Comment by bobh66** [comment]
One other data point which I think is already covered, we always retrieve all of our external dependencies and store them in a local repo and we use the local repo for all references/defaults.  That ensures we have local copies of everything for those rare occasions when a remote repo is unavailable, since we can't have a failure caused by remote artifacts being unavailable.  For packaging and distribution purposes we want to specify a default repo to the Crossplane instance and have all of our packages and dependencies with no repo so everything gets pulled from the default.  That way different deployments can use different local repos as needed.

**Comment by bobh66** [comment]
@phisco we haven't switched to the Upbound provider yet so it hasn't been a problem.  I do think the package manager should be able to fall back to the default registry if it can't find the fully qualified dependency, especially when you consider air-gapped systems that will never be able to go to the internet.  Or we need a way to generate a package that contains all referenced/nested dependencies in a single artifact.

**Comment by jbw976** [comment]
> But we think we should continue with our assumption that the lock file only takes into consideration exactly what has been specified, and not runtime configuration such as the default registry, which might change over time.

OK @phisco thanks for the spelunking to find that decision in the commit history. Sounds like we should continue to honor it, so I'll close this issue.

One thing we may still want to do in https://github.com/crossplane/docs/issues/680 is provide some guidance for how to clean up and get out of this scenario if a user gets into it accidentally.
